packages/by-name/kata/kata-runtime/0020-genpolicy-support-dynamic-annotations.patch

```diff
@@ -0,0 +1,149 @@
+--- a/src/tools/genpolicy/genpolicy-settings.json
++++ b/src/tools/genpolicy/genpolicy-settings.json
+@@ -309,7 +309,10 @@
+             "CAP_PERFMON",
+             "CAP_BPF",
+             "CAP_CHECKPOINT_RESTORE"
+-        ]
++        ],
++        "dynamic_annotations": {
++            "^cdi\\.k8s\\.io\\/vfio[0-9]{2}$": "^nvidia.com/gpu=[0-9]+$"
++        }
+     },
+     "kata_config": {
+         "confidential_guest": false,
+@@ -333,7 +336,8 @@
+                 "^AZURE_CLIENT_ID=[A-Fa-f0-9-]*$",
+                 "^AZURE_TENANT_ID=[A-Fa-f0-9-]*$",
+                 "^AZURE_FEDERATED_TOKEN_FILE=/var/run/secrets/azure/tokens/azure-identity-token$",
+-                "^AZURE_AUTHORITY_HOST=https://login\\.microsoftonline\\.com/$"
++                "^AZURE_AUTHORITY_HOST=https://login\\.microsoftonline\\.com/$",
++                "^PCI_RESOURCE_NVIDIA_COM.*=[a-fA-F0-9:.-]*$"
+             ]
+         },
+         "CopyFileRequest": [
+diff --git a/src/tools/genpolicy/rules.rego b/src/tools/genpolicy/rules.rego
+index 43cb19a56fe8ea5833708f0639c9e85ddd884cb3..271df2aebe05bd4bbd7aa396be24fb6fee0668bf 100644
+--- a/src/tools/genpolicy/rules.rego
++++ b/src/tools/genpolicy/rules.rego
+@@ -199,26 +199,31 @@ allow_anno(p_oci, i_oci) {
+ }
+ allow_anno(p_oci, i_oci) {
+     print("allow_anno 2: p Annotations =", p_oci.Annotations)
++    p_dynamic_annotations := policy_data.common.dynamic_annotations
++    print("allow_anno 2: p Dynamic Annotations =", p_dynamic_annotations)
++
++    i_annotations := i_oci.Annotations
+     print("allow_anno 2: i Annotations =", i_oci.Annotations)
+ 
+-    i_keys := object.keys(i_oci.Annotations)
++    i_keys := object.keys(i_annotations)
+     print("allow_anno 2: i keys =", i_keys)
+ 
+     every i_key in i_keys {
+-        allow_anno_key(i_key, p_oci)
++        allow_anno_key(i_key, p_oci, p_dynamic_annotations)
++        allow_dynamic_anno_value(i_key, i_annotations, p_dynamic_annotations)
+     }
+ 
+     print("allow_anno 2: true")
+ }
+ 
+-allow_anno_key(i_key, p_oci) {
++allow_anno_key(i_key, p_oci, p_dynamic_annotations) {
+     print("allow_anno_key 1: i key =", i_key)
+ 
+     startswith(i_key, "io.kubernetes.cri.")
+ 
+     print("allow_anno_key 1: true")
+ }
+-allow_anno_key(i_key, p_oci) {
++allow_anno_key(i_key, p_oci, p_dynamic_annotations) {
+     print("allow_anno_key 2: i key =", i_key)
+ 
+     some p_key, _ in p_oci.Annotations
+@@ -227,6 +232,49 @@ allow_anno_key(i_key, p_oci) {
+     print("allow_anno_key 2: true")
+ }
+ 
++allow_anno_key(i_key, p_oci, p_dynamic_annotations) {
++    print("allow_anno_key 3: i key =", i_key)
++
++    some p_key, _ in p_dynamic_annotations
++    regex.match(p_key, i_key)
++
++    print("allow_anno_key 3: true")
++}
++
++
++# Account for containers without dynamic annotations
++# at all..
++allow_dynamic_anno_value(i_key, i_annotations, p_dynamic_annotations) {
++    print("allow_dynamic_anno_value 1: i key =", i_key)
++
++    not p_dynamic_annotations
```

Comment:
This is more or less included in `allow_dynamic_anno_value 2`.

Reply:
In my testing, it seemed that `every` can't work with `p_dynamic_annotations` being `undefined`. Hence I had to add this.